platform/linux-generic/odp_crypto_openssl.c

```diff
@@ -1119,70 +1120,72 @@ static int process_cipher_param(odp_crypto_generic_session_t *session,
 	return 0;
 }
 
-static
-odp_crypto_alg_err_t cipher_encrypt_bits(odp_packet_t pkt,
-					 const odp_crypto_packet_op_param_t
-							*param,
-					 odp_crypto_generic_session_t *session)
+static odp_crypto_alg_err_t cipher_encrypt_bytes(odp_packet_t pkt,
+						 const odp_crypto_packet_op_param_t *param,
+						 odp_crypto_generic_session_t *session)
 {
 	EVP_CIPHER_CTX *ctx = local.cipher_ctx[session->idx];
 	int dummy_len = 0;
 	int cipher_len;
-	uint32_t in_len = (param->cipher_range.length + 7) / 8;
+	uint32_t in_len = param->cipher_range.length;
+	uint32_t offset = param->cipher_range.offset;
 	uint8_t data[in_len];
 	int ret;
-	uint32_t offset;
-
-	/* Range offset is in bits in bit mode but must be divisible by 8. */
-	offset = param->cipher_range.offset / 8;
 
 	EVP_EncryptInit_ex(ctx, NULL, NULL, NULL, param->cipher_iv_ptr);
-
 	odp_packet_copy_to_mem(pkt, offset, in_len, data);
-
 	EVP_EncryptUpdate(ctx, data, &cipher_len, data, in_len);
-
 	ret = EVP_EncryptFinal_ex(ctx, data + cipher_len, &dummy_len);
 	cipher_len += dummy_len;
-
 	odp_packet_copy_from_mem(pkt, offset, in_len, data);
-
 	return ret <= 0 ? ODP_CRYPTO_ALG_ERR_DATA_SIZE :
 			  ODP_CRYPTO_ALG_ERR_NONE;
 }
 
-static
-odp_crypto_alg_err_t cipher_decrypt_bits(odp_packet_t pkt,
-					 const odp_crypto_packet_op_param_t
-							*param,
-					 odp_crypto_generic_session_t *session)
+static odp_crypto_alg_err_t cipher_decrypt_bytes(odp_packet_t pkt,
+						 const odp_crypto_packet_op_param_t *param,
+						 odp_crypto_generic_session_t *session)
 {
 	EVP_CIPHER_CTX *ctx = local.cipher_ctx[session->idx];
 	int dummy_len = 0;
 	int cipher_len;
-	uint32_t in_len = (param->cipher_range.length + 7) / 8;
+	uint32_t in_len = param->cipher_range.length;
+	uint32_t offset = param->cipher_range.offset;
 	uint8_t data[in_len];
 	int ret;
-	uint32_t offset;
-
-	/* Range offset is in bits in bit mode but must be divisible by 8. */
-	offset = param->cipher_range.offset / 8;
 
 	EVP_DecryptInit_ex(ctx, NULL, NULL, NULL, param->cipher_iv_ptr);
-
 	odp_packet_copy_to_mem(pkt, offset, in_len, data);
-
 	EVP_DecryptUpdate(ctx, data, &cipher_len, data, in_len);
-
 	ret = EVP_DecryptFinal_ex(ctx, data + cipher_len, &dummy_len);
 	cipher_len += dummy_len;
-
 	odp_packet_copy_from_mem(pkt, offset, in_len, data);
-
 	return ret <= 0 ? ODP_CRYPTO_ALG_ERR_DATA_SIZE :
 			  ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static odp_crypto_alg_err_t cipher_encrypt_bits(odp_packet_t pkt,
+						const odp_crypto_packet_op_param_t *param,
+						odp_crypto_generic_session_t *session)
+{
+	odp_crypto_packet_op_param_t new_param = *param;
```

Comment:
Can we avoid this copy by passing offset & length as separate args to 'cipher_encrypt_bytes'?

Reply:
Since the prototype of the generic cipher function cannot be easily changed, that would require moving the body of cipher_encrypt_bytes() into a separate function and leaving cipher_encrypt() bytes to be a simpler wrapper. And the same for the decrypt function.

I rather wanted to keep the change simpler. Copying the parameter struct should not be that costly anyway. A better optimization would be to get rid of the packet data copy that occurs in cipher_encrypt_bytes by using internal_encrypt() that can handle also segmented packets without copying. But I did not want to do such a change yet in this PR which is about the bit mode semantics change.

Also, supporting OOP without the separate memory copying step would probably be a good optimization but that would require changing the prototype of the generic cipher & auth functions (and would also otherwise be a bit bigger change). Because of this, I did not want to start changing the generic cipher & auth function prototypes now.

Reply:
Understood. Thanks for the explanation.